actix-router/src/quoter.rs

```diff
@@ -1,132 +1,87 @@
+            if let (prev, [b'%', p1, p2, rem @ ..]) = val.split_at(i) {
+                if let Some(ch) = hex_pair_to_char(*p1, *p2)
+                    // ignore protected ascii bytes
+                    .filter(|&ch| !(ch < 128 && bit_at(&self.protected_table, ch)))
+                {
+                    *val = rem;
+                    return Some((prev, ch));
+                }
+            }
         }
 
-        quoter
+        None
     }
 
-    /// Decodes safe percent-encoded sequences from `val`.
-    ///
-    /// Returns `None` when no modification to the original byte string was required.
+    /// Partially percent-decodes the given bytes.
+    /// Escape sequences of the protected set are *not* decoded.
     ///
-    /// Non-ASCII bytes are accepted as valid input.
+    /// Returns `None` when no modification to the original bytes was required.
     ///
-    /// Behavior for invalid/incomplete percent-encoding sequences is unspecified and may include
-    /// removing the invalid sequence from the output or passing it as-is.
+    /// Invalid/incomplete percent-encoding sequences are passed unmodified.
     pub fn requote(&self, val: &[u8]) -> Option<Vec<u8>> {
-        let mut has_pct = 0;
-        let mut pct = [b'%', 0, 0];
-        let mut idx = 0;
-        let mut cloned: Option<Vec<u8>> = None;
-
-        let len = val.len();
-
-        while idx < len {
-            let ch = val[idx];
-
-            if has_pct != 0 {
-                pct[has_pct] = val[idx];
-                has_pct += 1;
-
-                if has_pct == 3 {
-                    has_pct = 0;
-                    let buf = cloned.as_mut().unwrap();
-
-                    if let Some(ch) = hex_pair_to_char(pct[1], pct[2]) {
-                        if ch < 128 {
-                            if bit_at(&self.protected_table, ch) {
-                                buf.extend_from_slice(&pct);
-                                idx += 1;
-                                continue;
-                            }
-
-                            if bit_at(&self.safe_table, ch) {
-                                buf.push(ch);
-                                idx += 1;
-                                continue;
-                            }
-                        }
-
-                        buf.push(ch);
-                    } else {
-                        buf.extend_from_slice(&pct[..]);
-                    }
-                }
-            } else if ch == b'%' {
-                has_pct = 1;
+        let mut remaining = val;
 
-                if cloned.is_none() {
-                    let mut c = Vec::with_capacity(len);
-                    c.extend_from_slice(&val[..idx]);
-                    cloned = Some(c);
-                }
-            } else if let Some(ref mut cloned) = cloned {
-                cloned.push(ch)
-            }
+        let (prev, ch) = self.decode_next(&mut remaining)?;
+        let mut buf = Vec::<u8>::with_capacity(val.len());
```

Comment:
At this line, we know there are replacements right? So in the end there will be more added to the vec than `val.len()`. Shall we just add a magic number of extra capacity to account for it?